src/gba/vfame.c

```diff
@@ -0,0 +1,298 @@
+	if (romSize != 0x400000) {
+		return false;
+	}
+	if (address < 0x800000) {
+		return false;
+	}
+	if (address >= 0x800000 + romSize) {
+		return false;
+	}
+	return true;
+}
+
+// Looks like only 16-bit reads are done by games but others are possible...
+uint32_t GBAVFameGetPatternValue(uint32_t address, int bits) {
+	switch (bits) {
+	case 8:
+		if (address & 1) {
+			return _getPatternValue(address) & 0xFF;
+		} else {
+			return (_getPatternValue(address) & 0xFF00) >> 8;
+		}
+	case 16:
+		return _getPatternValue(address);
+	case 32:
+		return (_getPatternValue(address) << 2) + _getPatternValue(address + 2);
+	}
+	return 0;
+}
+
+// when you read from a ROM location outside the actual ROM data or its mirror, it returns a value based on some 16-bit transformation of the address
+// which the game relies on to run
+static uint32_t _getPatternValue(uint32_t addr) {
+	addr &= 0x1FFFFF;
+	uint32_t value = 0;
+	switch (addr & 0x1F0000) {
+	case 0x000000:
+	case 0x010000:
+		value = (addr >> 1) & 0xFFFF;
+		break;
+	case 0x020000:
+		value = addr & 0xFFFF;
+		break;
+	case 0x030000:
+		value = (addr & 0xFFFF) + 1;
+		break;
+	case 0x040000:
+		value = 0xFFFF - (addr & 0xFFFF);
+		break;
+	case 0x050000:
+		value = (0xFFFF - (addr & 0xFFFF)) - 1;
+		break;
+	case 0x060000:
+		value = (addr & 0xFFFF) ^ 0xAAAA;
+		break;
+	case 0x070000:
+		value = ((addr & 0xFFFF) ^ 0xAAAA) + 1;
+		break;
+	case 0x080000:
+		value = (addr & 0xFFFF) ^ 0x5555;
+		break;
+	case 0x090000:
+		value = ((addr & 0xFFFF) ^ 0x5555) - 1;
+		break;
+	case 0x0A0000:
+	case 0x0B0000:
+		value = _patternRightShift2(addr);
+		break;
+	case 0x0C0000:
+	case 0x0D0000:
+		value = 0xFFFF - _patternRightShift2(addr);
+		break;
+	case 0x0E0000:
+	case 0x0F0000:
+		value = _patternRightShift2(addr) ^ 0xAAAA;
+		break;
+	case 0x100000:
+	case 0x110000:
+		value = _patternRightShift2(addr) ^ 0x5555;
+		break;
+	case 0x120000:
+		value = 0xFFFF - ((addr & 0xFFFF) >> 1);
+		break;
+	case 0x130000:
+		value = 0xFFFF - ((addr & 0xFFFF) >> 1) - 0x8000;
+		break;
+	case 0x140000:
+	case 0x150000:
+		value = ((addr >> 1) & 0xFFFF) ^ 0xAAAA;
+		break;
+	case 0x160000:
+	case 0x170000:
+		value = ((addr >> 1) & 0xFFFF) ^ 0x5555;
+		break;
+	case 0x180000:
+	case 0x190000:
+		value = ((addr >> 1) & 0xFFFF) ^ 0xF0F0;
+		break;
+	case 0x1A0000:
+	case 0x1B0000:
+		value = ((addr >> 1) & 0xFFFF) ^ 0x0F0F;
+		break;
+	case 0x1C0000:
+	case 0x1D0000:
+		value = ((addr >> 1) & 0xFFFF) ^ 0xFF00;
+		break;
+	case 0x1E0000:
+	case 0x1F0000:
+		value = ((addr >> 1) & 0xFFFF) ^ 0x00FF;
+		break;
+	}
+
+	return value & 0xFFFF;
+}
+
+static uint32_t _patternRightShift2(uint32_t addr) {
+	uint32_t value = addr & 0xFFFF;
+	value >>= 2;
+	value += (addr & 3) == 2 ? 0x8000 : 0;
+	value += (addr & 0x10000) ? 0x4000 : 0;
+	return value;
+}
+
+void GBAVFameSramWrite(struct GBAVFameCart* cart, uint32_t address, int8_t value, uint8_t* sramData) {
```

Comment:
It looks like this won't come up in any Vast Fame games, but value should be uint8_t, or else a game can set the mode to -1, which is your reserved value.